### PR TITLE
Enable citus.defer_drop_after_shard_move by default

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
DESCRIPTION: Enables citus.defer_drop_after_shard_move by default

Some issues to consider:

https://github.com/citusdata/citus-enterprise/issues/585
https://github.com/citusdata/citus/issues/4954

### CODE REVIEW

File: src/backend/distributed/shared_library_init.c [hunk 1]
```diff
@@ -654,7 +654,7 @@ RegisterCitusConfigVariables(void)
 					 "the background worker moves on. When set to -1 this background "
 					 "process is skipped."),
 		&DeferShardDeleteInterval,
-		-1, -1, 7 * 24 * 3600 * 1000,
+		60000, -1, 7 * 24 * 3600 * 1000,
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/shared_library_init.c:657]: I think we can make the default 15 seconds because it might be confusing for users to see that there are still old shards even though they did a rebalance.

Comment (marcocitus) [reply src/backend/distributed/shared_library_init.c:657]: I have found a minute to be tediously long when testing things.

Comment (SaitTalhaNisanci) [reply src/backend/distributed/shared_library_init.c:657]: Yeah a minute is long, is 15 seconds also long?

Comment (marcocitus) [reply src/backend/distributed/shared_library_init.c:657]: I'd be ok with 15 seconds. I think shorter might be problematic when things are failing.

File: src/backend/distributed/operations/repair_shards.c [hunk 2]
```diff
@@ -300,6 +300,14 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
 	Oid relationId = RelationIdForShard(shardId);
 	ErrorIfMoveCitusLocalTable(relationId);
 
+	bool waitForCleanupLock = false;
+
+	/*
+	 * We try to drop marked shards so that we won't unnecessarily error
+	 * for old shard placements.
+	 */
+	TryDropMarkedShards(waitForCleanupLock);
```
Comment (marcocitus) [inline src/backend/distributed/operations/repair_shards.c:309]: should we wait for the cleanup lock here?

File: src/test/regress/spec/isolation_blocking_move_multi_shard_commands_on_mx.spec [hunk 3]
```diff
@@ -60,7 +61,8 @@ step "s1-begin"
 
 step "s1-move-placement"
 {
-    	SELECT master_move_shard_placement(get_shard_id_for_distribution_column, 'localhost', 57637, 'localhost', 57638, shard_transfer_mode:='block_writes') FROM selected_shard;
+  SET citus.defer_drop_after_shard_move to off;
```
Comment (JelteF) [inline src/test/regress/spec/isolation_blocking_move_multi_shard_commands_on_mx.spec:64]: Why is this needed? I'd expect it not to be necessary with the `shardstate != 4` changes.

File: src/test/regress/spec/isolation_blocking_move_single_shard_commands.spec [hunk 4]
```diff
@@ -31,7 +32,8 @@ step "s1-begin"
 
 step "s1-move-placement"
 {
-    	SELECT master_move_shard_placement((SELECT * FROM selected_shard), 'localhost', 57637, 'localhost', 57638, shard_transfer_mode:='block_writes');
+    SET citus.defer_drop_after_shard_move to off;
```
Comment (JelteF) [inline src/test/regress/spec/isolation_blocking_move_single_shard_commands.spec:35]: Why is this needed? I'd expect it not to be necessary with the shardstate != 4 changes.

**Review by JelteF — CHANGES_REQUESTED** [review]

File: src/backend/distributed/operations/repair_shards.c (hunk 2 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:309]: done

File: src/backend/distributed/operations/repair_shards.c [hunk 5]
```diff
@@ -1015,9 +1023,22 @@ EnsureShardCanBeCopied(int64 shardId, const char *sourceNodeName, int32 sourceNo
 																 targetNodePort);
 	if (targetPlacement != NULL)
 	{
-		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-						errmsg("shard " INT64_FORMAT " already exists in the target node",
-							   shardId)));
+		if (targetPlacement->shardState == SHARD_STATE_TO_DELETE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
```
Comment (JelteF) [inline src/backend/distributed/operations/repair_shards.c:1028]: Can you add a test that covers this line? (if it's not too hard)

File: src/test/regress/expected/isolation_rebalancer_deferred_drop.out [hunk 6]
```diff
@@ -34,20 +33,16 @@ step s1-begin:
...
 step s2-drop-marked-shards:
     SET client_min_messages to DEBUG1;
     SELECT public.master_defer_delete_shards();
-
-master_defer_delete_shards
-
-0
-step s1-drop-marked-shards:
+ <waiting ...>
+step s1-drop-marked-shards:
```
Comment (JelteF) [inline src/test/regress/expected/isolation_rebalancer_deferred_drop.out:45]: This seems like a change that should not have happened. Is this happening consistently? Given we set `citus.defer_shard_delete_interval=-1` in the default test environment, I don't think it would be caused by the background worker.

File: src/test/regress/expected/shard_move_deferred_delete.out [hunk 7]
```diff
@@ -169,13 +169,12 @@ SELECT master_move_shard_placement(20000001, 'localhost', :worker_2_port, 'local
...
 
 ROLLBACK;
--- we expect shard xxxxx to be on both of the workers
 SELECT run_command_on_workers($cmd$
     SELECT count(*) FROM pg_class WHERE relname = 't1_20000000';
 $cmd$);
  run_command_on_workers
 ---------------------------------------------------------------------
- (localhost,57637,t,1)
+ (localhost,57637,t,0)
```
Comment (JelteF) [inline src/test/regress/expected/shard_move_deferred_delete.out:177]: Why did this change?

Comment (SaitTalhaNisanci) [reply src/test/regress/expected/shard_move_deferred_delete.out:177]: Because when moving `2000000`  we first drop the old shard placements now in the call of `master_move_shard_placement` so the old `t1_20000000` is deleted.

File: src/backend/distributed/operations/repair_shards.c (hunk 5 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:1028]: I was trying to add a test, it wasn't super easy so I manually tested but it should be possible to add a test for this line so I will try

Comment (SaitTalhaNisanci) [reply src/backend/distributed/operations/repair_shards.c:1067]: added

**Review by JelteF — APPROVED** [review]
